### PR TITLE
OAuth Application Resource

### LINKED ISSUE #56: create an `okta_app` resource to configure new applications in okta
Will need to create an OpenID Connect Okta Application to tie back to rise.com

### PR DESCRIPTION
**Pull request by quantumew**
Implement the OpenID OAuth Application resource. Fixes #56. 

General thoughts:
* Switched to develop branch of the Okta SDK. This was the only way to get a working Application API and even still ListApplications is broken.

* Please closely code review the implementation around grant_types, application_type, and response_types. I took the approach that allows honing of configuration by specifying all three pieces of information but also have some logic that defaults response_types and grant_types based on configuration. Another approach would be to have 4 separate resources for these 4 OAuth types. This seemed excessive in this case. Especially as we add other Apps. 

* Starting the habit of unit testing utility or pure Go functions. Should I pull in a BDD library (Gingko for instance) for this? I personally like that style of testing.

* Slowly but surely adding features to reduce test code duplication. I have an idea i started elsewhere of a test framework that can set you up a default test with checks based on the config you supply it. Hoping to finish that one of these weekends.

* I see no good way to introspect during the Validation lifecycle step, making conditional logic difficult. For instance, if field A is "foo" then field B is required, else it is not, etc... Going to ask about this in the Google group.

### CODE REVIEW

File: okta/resource_user.go [hunk 1]
```diff
@@ -268,7 +268,7 @@ func resourceUserRead(d *schema.ResourceData, m interface{}) error {
 		return nil
 	}
 
-	user, _, err := client.User.GetUser(d.Id(), nil)
+	user, _, err := client.User.GetUser(d.Id())
```
Comment (quantumew) [inline okta/resource_user.go]: API change in Okta SDK, hence all of these changes

File: okta/resource_oauth_app_test.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	"testing"
+
+	articulateOkta "github.com/articulate/oktasdk-go/okta"
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/okta/okta-sdk-golang/okta"
+	"github.com/okta/okta-sdk-golang/okta/query"
+)
+
+func deleteOAuthApps(artClient *articulateOkta.Client, client *okta.Client) error {
```
Comment (PurrBiscuit) [inline okta/resource_oauth_app_test.go]: is the `artClient` parameter needed?  it doesn't look like it's being used in this delete func.  We could get rid of the `github.com/articulate/oktasdk-go/okta` import above if we can get rid of that there.

File: okta/resource_oauth_app.go [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Name of resource.",
+			},
+			"signon_mode": &schema.Schema{
```
Comment (PurrBiscuit) [inline okta/resource_oauth_app.go]: nitpicky incoming - I think this should be `"sign_on_mode"`

File: okta/resource_oauth_app.go [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+func resourceOAuthAppRead(d *schema.ResourceData, m interface{}) error {
+	app, err := fetchApp(d, m)
+
+	if err != nil {
+		return err
+	}
+
+	d.Set("name", app.Name)
+	d.Set("status", app.Status)
+	d.Set("signon_mode", app.SignOnMode)
```
Comment (PurrBiscuit) [inline okta/resource_oauth_app.go]: same goes for here with my nitpicky comment above about `"sign_on_mode"`

File: okta/resource_oauth_app_test.go (hunk 2 above)
Comment (quantumew) [reply okta/resource_oauth_app_test.go]: The only reason why it takes both is for a predictable interface for setting up sweeping

https://github.com/articulate/terraform-provider-okta/blob/master/okta/provider_sweeper_test.go#L27

With that said there are ways to achieve both of our ends so I will change this!

Comment (quantumew) [reply okta/resource_oauth_app_test.go]: kabooom f643a24347d9cffe37009c4369329c94331c2634

**Pull request merged**